p2panda-store/src/lib.rs

```diff
@@ -1,7 +1,112 @@
 // SPDX-License-Identifier: AGPL-3.0-or-later
 
+#[cfg(feature = "memory")]
 pub mod memory_store;
-pub mod traits;
 
+#[cfg(feature = "memory")]
 pub use memory_store::MemoryStore;
-pub use traits::{LogStore, OperationStore, StoreError};
+
+use p2panda_core::{Body, Hash, Header, PublicKey};
+
+#[trait_variant::make(OperationStore: Send)]
+pub trait LocalOperationStore<LogId, Extensions> {
+    type Error;
+
+    /// Insert an operation.
+    ///
+    /// Returns `true` when the insert occurred, or `false` when the operation already existed and
+    /// no insertion occurred.
+    async fn insert_operation(
+        &mut self,
+        hash: Hash,
+        header: &Header<Extensions>,
+        body: Option<&Body>,
+        header_bytes: &[u8],
+        log_id: &LogId,
+    ) -> Result<bool, Self::Error>;
+
+    /// Get an operation.
+    async fn get_operation(
+        &self,
+        hash: Hash,
+    ) -> Result<Option<(Header<Extensions>, Option<Body>)>, Self::Error>;
```

Comment:
Maybe it would be nice to still get `Operation`s back from the store, even though the insert method now takes it's constituent parts?

Reply:
Pros:
- minor (but often occurring) optimisation that the hash doesn't need to be re-generated
- we will likely use `Operation` in other APIs so why not provide that already from the store

Cons:
- if point 2 above isn't actually true :sweat_smile:

Reply:
I've refactored all APIs which touch the store and realized we're so far fine with returning `(Header, Option<Body>)` instead of `Operation`. The only method where we are hashing and assuming `Operation` (for example in `validate_operation`), is `ingest_operation` (in `p2panda-engine`), but everything fits nicely together there as it seems and no step is redundant ✨